File: day5/part1/intcode.py
def get_modes(instructions, idx):
    params = []
    instruction = instructions[idx]

    mode3 = instruction // 10000
    instruction -= 10000*mode3

    mode2 = instruction // 1000
    instruction -= 1000*mode2

    mode1 = instruction // 100
    instruction -= 100*mode1

    opcode = instruction

    return opcode, [mode1, mode2, mode3]


def get_params(instructions, idx, opcode, modes):
    params = []
    if opcode == 1 or opcode == 2:
        if modes[0] == 0:
            params.append(instructions[idx+1])
        else:
            params.append(idx+1)
        if modes[1] == 0:
            params.append(instructions[idx+2])
        else:
            params.append(idx+2)
        if modes[2] == 0:
            params.append(instructions[idx+3])
        else:
            params.append(idx+3)

    elif opcode == 3 or opcode == 4:
        if modes[0] == 0:
            params.append(instructions[idx+1])
        else:
            params.append(idx+1)

    return params


def interpret(instructions, idx):
    opcode, modes = get_modes(instructions, idx)
    params = get_params(instructions, idx, opcode, modes)

    print("Raw instruction:", instructions[idx])
    print("Idx:", idx)
    print("Opcode:", opcode)
    print("Params:", params)
    print("Modes:", modes)

    if opcode == 1:
        print("Sum: ", instructions[params[0]], "+", instructions[params[1]])
        instructions[params[2]] = instructions[params[0]] + instructions[params[1]]
        return idx+4

    elif opcode == 2:
        print("Mult: ", instructions[params[0]], "*", instructions[params[1]])
        instructions[params[2]] = instructions[params[0]] * instructions[params[1]]
        return idx+4

    elif opcode == 3:
        instructions[params[0]] = int(input())
        return idx+2

    elif opcode == 4:
        print("Opcode 4: ", instructions[params[0]])
        return idx+2

    elif opcode == 99:
        return -1

    else:
        print("Unexpected opcode ", opcode)
        print(instructions)
        exit(1)



if __name__ == "__main__":

    with open("intcode_input.txt") as fp:
        for line in fp:
            instructions = [int(a) for a in line.strip().split(',')]

    idx = 0

    while idx != -1:
        idx = interpret(instructions, idx)

    print("FINISHED")
    # print(instructions)

File: day5/part1/intcode.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type instruction struct {
	opcode  int
	nparams int
	name    string
}

var instructionTypes = map[int]instruction{
	1:  {1, 3, "sum"},
	2:  {2, 3, "mult"},
	3:  {3, 1, "input"},
	4:  {4, 1, "output"},
	5:  {5, 2, "jump-if-true"},
	6:  {6, 2, "jump-if-false"},
	7:  {7, 3, "less than"},
	8:  {8, 3, "equals"},
	99: {99, 0, "halt"},
}

func exec(instructions []int, IP int) int {
	params := getParams(instructions, IP)
	modes := getModes(instructions, IP)
}

// Run a sequence of instructions
func Run(instructions []int) []int {
	newIP := 0
	IP := 1

	for IP != newIP {
		IP = newIP
		newIP := exec(IP, instructions)
	}

	for opcode := instructions[idx]; opcode != 99; opcode = instructions[idx] {
		a := instructions[idx+1]
		b := instructions[idx+2]
		i := instructions[idx+3]
		if opcode == 1 {
			instructions[i] = instructions[a] + instructions[b]
		} else if opcode == 2 {
			instructions[i] = instructions[a] * instructions[b]
		} else {
			fmt.Printf("Unexpected opcode %d\n", opcode)
			os.Exit(1)
		}

		idx += 4
	}

	return instructions
}

// LoadRaw reads an intcode program from stdin
// and saves it to strip and the instructions to
// an array of instrucitons
func LoadRaw(rawText string) []int {
	var strip []int

	line := strings.Split(rawText, ",")
	for i := 0; i < len(line); i++ {
		intnum, _ := strconv.Atoi(line[i])
		strip = append(strip, intnum)
	}

	return strip
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()

	strip := LoadRaw(scanner.Text())

	endStrip := execStrip(instructions)

	fmt.Println(endStrip[0])
}
